File: app/src/main/java/com/nodoubles/app/Adapters/RosterAdapter.java
package com.nodoubles.app.Adapters;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.res.Resources;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.nodoubles.app.App;
import com.nodoubles.app.EditFighterActivity;
import com.nodoubles.app.Models.Fighter;
import com.nodoubles.app.R;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.Locale;


public class RosterAdapter extends RecyclerView.Adapter<RosterAdapter.CustomViewHolder> {

    public int getItemCount() {return fighters.size();}
    private Context context;
    private ArrayList<Fighter> fighters;
    private Resources res;

    public RosterAdapter(Context context, ArrayList<Fighter> fighters){
        this.fighters = fighters;
        this.context = context;
        res = context.getResources();

    }

    @Override
    public CustomViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.fighter_card, parent, false);
        return new CustomViewHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull CustomViewHolder h, int i) {
        final Fighter fighter = fighters.get(i);
        h.root.setOnClickListener(new View.OnClickListener(){
            public void onClick(View view) {
                if(App.Globals.INSTANCE.isAdmin()){
                    Intent i = new Intent(context, EditFighterActivity.class);
                    i.putExtra("fighter", fighter.getId());
                    context.startActivity(i);
                }
            }
        });
        h.root.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View view) {
                if(App.Globals.INSTANCE.isAdmin()) {
                    AlertDialog.Builder alert = new AlertDialog.Builder(context);
                    alert.setTitle(R.string.delete_q);
                    alert.setMessage(R.string.delete_warning);
                    alert.setPositiveButton(R.string.DELETE, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            //deleteFighter(fighter);
                        }
                    });
                    alert.setNegativeButton(R.string.CANCEL, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            dialog.dismiss();
                        }
                    });
                    alert.show();
                }
                return false;
            }
        });
        h.name.setText(fighter.getFullName()) ;
        h.stats.setText(formatStats(fighter));
        String url = fighter.getPhotoURL();
        if(url != null && !url.equals(""))
            Picasso.get()
                    .load(url)
                    .placeholder(res.getDrawable(R.drawable.ic_sword_cross))
                    .error(res.getDrawable(R.drawable.ic_sword_cross))
                    .into(h.profile);
        else
            h.profile.setImageDrawable(res.getDrawable(R.drawable.ic_sword_cross));

    }

    private String formatStats(Fighter f){
        int w = f.getWins();
        int l = f.getLosses();
        int p = f.getTourneyScore();
        return String.format(Locale.CANADA,
                "%s%3d%n%s%3d%n%s%3d",
                context.getString(R.string.wins_abrev)  ,w,
                context.getString(R.string.losses_abrev),l,
                context.getString(R.string.points_abrev),p);
    }

//    private void deleteFighter (final Fighter fighter){
//        RealmConfiguration config = SyncUser.current()
//                .createConfiguration(App.Globals.INSTANCE_ADDRESS + App.Globals.REALM)
//                .build();
//        Realm.getInstance(config).executeTransaction(new Realm.Transaction() {
//            @Override
//            public void execute(Realm realm) {
//                ArrayList<Fight> res2 = realm
//                        .where(Fight.class)
//                        .equalTo("fighter1.id", fighter.getId())
//                        .or()
//                        .equalTo("fighter2.id", fighter.getId())
//                        .findAll();
//                res2.deleteAllFromRealm();
//                ArrayList<Fighter> res = realm
//                        .where(Fighter.class)
//                        .equalTo("id", fighter.getId())
//                        .findAll();
//                res.deleteAllFromRealm();
//            }
//        });
//    }

    class CustomViewHolder extends RecyclerView.ViewHolder {

        View root;
        ImageView profile;
        TextView name, stats;

        CustomViewHolder(View v) {
            super(v);
            root = v.getRootView();
            profile = v.findViewById(R.id.fighter_img);
            name = v.findViewById(R.id.fighter_name);
            stats = v.findViewById(R.id.fighter_stats);
        }
    }

}

File: app/src/main/java/com/nodoubles/app/Adapters/RosterAdapter.kt
package com.nodoubles.app.Adapters

import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.content.res.Resources
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener

import com.nodoubles.app.App
import com.nodoubles.app.EditFighterActivity
import com.nodoubles.app.Models.Fight
import com.nodoubles.app.Models.Fighter
import com.nodoubles.app.R
import com.squareup.picasso.Picasso

import java.util.ArrayList
import java.util.Locale


class RosterAdapter(private val context: Context, private val fighters: ArrayList<Fighter>) : RecyclerView.Adapter<RosterAdapter.CustomViewHolder>() {
    private val res: Resources = context.resources

    override fun getItemCount(): Int {
        return fighters.size
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CustomViewHolder {
        val v = LayoutInflater.from(parent.context)
                .inflate(R.layout.fighter_card, parent, false)
        return CustomViewHolder(v)
    }

    override fun onBindViewHolder(h: CustomViewHolder, i: Int) {
        val fighter = fighters[i]
        h.root.setOnClickListener {
            if (App.Globals.isAdmin) {
                val i = Intent(context, EditFighterActivity::class.java)
                i.putExtra("fighter", fighter.id)
                context.startActivity(i)
            }
        }
        h.root.setOnLongClickListener {
            if (App.Globals.isAdmin) {
                val alert = AlertDialog.Builder(context)
                alert.setTitle(R.string.delete_q)
                alert.setMessage(R.string.delete_warning)
                alert.setPositiveButton(R.string.DELETE) { _, _ -> deleteFighter(fighter) }
                alert.setNegativeButton(R.string.CANCEL) { dialog, _ -> dialog.dismiss() }
                alert.show()
            }
            false
        }
        h.name.text = fighter.getFullName()
        h.stats.text = formatStats(fighter)
        val url = fighter.photoURL
        if (url != null && url != "")
            Picasso.get()
                    .load(url)
                    .placeholder(res.getDrawable(R.drawable.ic_sword_cross))
                    .error(res.getDrawable(R.drawable.ic_sword_cross))
                    .into(h.profile)
        else
            h.profile.setImageDrawable(res.getDrawable(R.drawable.ic_sword_cross))

    }

    private fun formatStats(f: Fighter): String {
        val w = f.wins
        val l = f.losses
        val p = f.tourneyScore
        return String.format(Locale.CANADA,
                "%s%3d%n%s%3d%n%s%3d",
                context.getString(R.string.wins_abrev), w,
                context.getString(R.string.losses_abrev), l,
                context.getString(R.string.points_abrev), p)
    }

    private fun deleteFighter(fighter: Fighter) {
        var fightsDeleted = 0

        val ref = App.Globals.db.reference.child("fights").child(App.Globals.TourneyID.toString())
        val listener = object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                for (postSnapshot in dataSnapshot.children) {
                    val fight = postSnapshot.getValue(Fight::class.java)
                    if(fight?.status != Fight.STATUS_FINISHED
                            && (fight?.fighter1?.id == fighter.id || fight?.fighter2?.id == fighter.id)){
                        deleteFight(fight)
                        ++fightsDeleted
                    }
                }

                App.Globals.db
                        .reference
                        .child("fighters")
                        .child(App.Globals.TourneyID.toString())
                        .child(fighter.id.toString())
                        .setValue(null)

                if(fightsDeleted != 0)
                    Toast.makeText(context, "$fightsDeleted fights deleted.",
                        Toast.LENGTH_SHORT).show()

            }
            override fun onCancelled(databaseError: DatabaseError) {
                Toast.makeText(context, "Failed delete fighter.",
                        Toast.LENGTH_SHORT).show()
            }
        }

        ref.addValueEventListener(listener)
    }

    private fun deleteFight(fight: Fight) {
        App.Globals.db
                .reference
                .child("fights")
                .child(App.Globals.TourneyID.toString())
                .child(fight.id.toString())
                .setValue(null)
    }

    inner class CustomViewHolder(v: View) : RecyclerView.ViewHolder(v) {

        var root: View = v.rootView
        var profile: ImageView = v.findViewById(R.id.fighter_img)
        var name: TextView = v.findViewById(R.id.fighter_name)
        var stats: TextView = v.findViewById(R.id.fighter_stats)

    }

}
